docs/specification.md

```diff
@@ -110,9 +96,13 @@ These values are:
 # State model
 
 This chapter provides a high-level overview of the contract's state.
-Note that the exact storage type is implementation detail and is not specified here.
+Note that the exact storage type is in the implementation detail and is not specified here.
 
-**TODO**
+- The contract maintains a ledger of address and its balance (frozen and unfrozen tokens)
+- The contract manages a special role called "Administrator".
+- The contract stores a list of proposals with one of states: "ongoing", "rejected", or "accepted.
+- After a successful migration, the contract `migrated` state will be `True` and all the subsequent
+  operation will be fail with `CONTRACT_MIGRATED` and the updated contract address.
```

Comment:
I've updated all the changes listed in `TODO`, Ivan.

One thing that I'm not clear is the "State Model". I've added this list, as well as the 'Ledger' section below. Let me know if this is the same as what you have in mind.

Reply:
The googledoc lists "What is the frozen token ID" among "On-chain Information" that should be exposed. It should return the address of the associated FA2 contract, in our case it's `SELF`. There are several ways to make some data readable from the contract, the most common ways are:
* put it into the storage and specify how it can be found (e. g. by annotation);
* add a `view` entrypoint, that is an entrypoint that takes a callback contract.

The first option is quite complicated in our case because we don't know `SELF` address during origination. So the only way to put it into the storage is by having a dedicated entrypoint and calling it once after origination. For this reason it seems easier to have a view entrypoint instead.

So yes, it should be also an entrypoint, it should take `contract address` I suppose.

Reply:
> One thing that I'm not clear is the "State Model". I've added this list, as well as the 'Ledger' section below.

It looks good. I think it should be extended a bit because some logic is not described there. I think I'll do it myself a bit later today.

Reply:
> So yes, it should be also an entrypoint, it should take `contract address` I suppose.

We have `token_metadata_registry` like this, and I was wondering if it is the same thing as you describe or we need a new entrypoint similar to this, Ivan?
```
type TokenMetadataRegistryParam = ContractRef Address

data Parameter proposalMetadata
  = Token_metadata_registry TokenMetadataRegistryParam

Parameter (in Michelson)
(contract %token_metadata_registry address)

- Return contract address that holds token metadata.
- Token metadata MUST include the DAO metadata.
```

Reply:
`token_metadata_registry` is a slightly different thing that uses similar mechanics. In our case, `tokenAddress` getter returns `SELF`, but `token_metadata_registry` may return something different. In `stablecoin`, we store metadata in a separate contract to keep the main contract a bit smaller. Here we can do the same :shrug:

Reply:
I see, so in other word, do they serve the same purpose? If they do, maybe we want to replace `token_metadata_registry` with `token_address` and if not, I'll add a separate `token_address` entrypoint.

Reply:
@rinn7e not exactly the same, they are semantically different. We can't remove `token_metadata_registry` because it's required by FA2. So I think adding `token_address` is the only way to go.

Reply:
I've added this, let me know if it's good enough, Ivan.